### PR TITLE
RMLMapper Optimizations v2

### PR DESCRIPTION
**Pull request by imad-99**
Hi, 

As the old PR was not up to date with the new version, I made a new fork. Can you please merge this one in. Almost all changes are the same as in the earlier PR. If you have questions feel free to send me a mail: imadelfetouh@hotmail.com

Imad

### CODE REVIEW

File: src/main/java/be/ugent/rml/Executor.java [hunk 1]
```diff
@@ -3,26 +3,43 @@
...
+import java.util.stream.IntStream;
+import java.util.stream.Stream;
 
 public class Executor {
 
     private static final Logger logger = LoggerFactory.getLogger(Executor.class);
 
+    private Map<Pair<Term, Extractor>, MultiValuedMap<List<Object>, ProvenancedTerm>> cachedRecordMap = new HashMap<>();
+
+    public static class Vocabulary {
```
Comment (DylanVanAssche) [inline src/main/java/be/ugent/rml/Executor.java:37]: This is the RML vocabulary, it should not be here.

File: src/main/java/be/ugent/rml/Executor.java [hunk 2]
```diff
@@ -220,12 +237,17 @@ private List<PredicateObjectGraph> generatePredicateObjectGraphs(Mapping mapping
...
                 //check if need to apply a join condition
                 if (!pogMapping.getJoinConditions().isEmpty()) {
-                    objects = this.getIRIsWithConditions(record, pogMapping.getParentTriplesMap(), pogMapping.getJoinConditions());
-                    //this.generateTriples(subject, po.getPredicateGenerator(), objects, record, combinedGraphs);
+                    Map<String, Object[]> parameters = ((StaticMultipleRecordsFunctionExecutor) pogMapping.getJoinConditions().iterator().next()).getParameters();
+                    if (isSimpleJoin(pogMapping, parameters)) {
+                        objects = this.getIRIsWithSimpleJoin(record, pogMapping.getParentTriplesMap(), parameters);
+                    } else {
+                        objects = this.getIRIsWithConditions(record, pogMapping.getParentTriplesMap(), pogMapping.getJoinConditions());
+                        //this.generateTriples(subject, po.getPredicateGenerator(), objects, record, combinedGraphs);
```
Comment (DylanVanAssche) [inline src/main/java/be/ugent/rml/Executor.java:249]: Let's not copy commented out dead code ;) Remove this please.

File: src/main/java/be/ugent/rml/Executor.java [hunk 3]
```diff
@@ -237,6 +259,67 @@ private List<PredicateObjectGraph> generatePredicateObjectGraphs(Mapping mapping
...
+                .filter(a -> parameterName.equals(a[0]))
+                .map(a -> a[1])
+                .filter(Extractor.class::isInstance)
+                .map(Extractor.class::cast)
+                .findAny()
+                .orElse(null);
+    }
+
+    // Check if the 2 parameters exist
+    private boolean isSimpleJoin(PredicateObjectGraphMapping pogMapping, Map<String, Object[]> parameters) {
```
Comment (DylanVanAssche) [inline src/main/java/be/ugent/rml/Executor.java:306]: This function must be documented, what is a 'simple join'?
What are the exact criteria to be a 'simple join' or a normal one?

**Review by DylanVanAssche — CHANGES_REQUESTED** [review]
Please use a proper commit message, the current git commit message does describe any of the changes.
Ideally, there would be smaller commits, one-per-change, but I will allow it to use a one commit if each change is properly explained in the description. This is needed for a proper git history.

**Comment by DylanVanAssche** [comment]
> Hi Dylan, most of the changes are described in the old PR (https://github.com/RMLio/rmlmapper-java/pull/175). You also responded on those a year ago (also the simple join). As i said I made this new fork and PR as the old one was not up to date anymore, but most of the changes are the same as in the old PR.

Yes, but the explanation like 'what is a simple join' would need to be a comment inside the code. Just like the other changes. Otherwise this will be broken very quickly in the future as other developers will not be aware why it was written like you did to gain some performance improvements. I asked for this before but it has not happen yet.

File: src/main/java/be/ugent/rml/Executor.java (hunk 1 above)
Comment (imad-99) [reply src/main/java/be/ugent/rml/Executor.java:37]: Hi Dylan, is there already a class for RML vocabulary?

Comment (DylanVanAssche) [reply src/main/java/be/ugent/rml/Executor.java:37]: IRI namespaces go into https://github.com/RMLio/rmlmapper-java/blob/master/src/main/java/be/ugent/rml/NAMESPACES.java

After that, there's no real class for that. You just need to move the `EQUAL_URI` and use string concat. If you want to make it a separate class like here, please make a separate file `RML.java` in https://github.com/RMLio/rmlmapper-java/blob/master/src/main/java/be/ugent/rml/

**Comment by imad-99** [comment]
Hi Dylan, I've resolved those 3 points

File: src/main/java/be/ugent/rml/RML.java [hunk 4]
```diff
@@ -0,0 +1,7 @@
+package be.ugent.rml;
+
+public class RML {
+    public static final String EQUAL_URI = "http://example.com/idlab/function/equal";
```
[the remark this answers is not in the document]
Comment (imad-99) [reply src/main/java/be/ugent/rml/RML.java]: Fixed

[the remark this answers is not in the document]
Comment (DylanVanAssche) [reply src/main/java/be/ugent/rml/RML.java]: No, don't move `parent` and `child` please, they are from the RML vocabulary.
The `RML.java` file should not be removed, that file has` parent` and `child` PARAMETERS. Only the equal one is for `IDLABFN.java`.

[the remark this answers is not in the document]
Comment (imad-99) [reply src/main/java/be/ugent/rml/RML.java]: Sorry for my misunderstanding, now it's fixed

[the remark this answers is not in the document]
Comment (DylanVanAssche) [reply src/main/java/be/ugent/rml/RML.java]: No problem :) Anyway, we are there now! I will squash the commits and merge it internally once I'm back from vacation which is on Monday. Thank you for your contributions.

[the remark this answers is not in the document]
Comment (imad-99) [reply src/main/java/be/ugent/rml/RML.java]: Thanks! thank you also for your cooperation

**Comment by DylanVanAssche** [comment]
Merged into development.

**Pull request closed without merging**